### ISSUE TITLE
Trim duplicate objects

### ISSUE THREAD

**Issue by leesei** [Kozea/WeasyPrint#969]
I'm using WeasyPrint to create lucky draw tickets, using a jinja template with images (logos and stuffs).
It seems WeasyPrint will embed the image for each instance rather then reusing the same object, resulting in a bloated PDF file.

I tried [pts/pdfsizeopt: PDF file size optimizer](https://github.com/pts/pdfsizeopt) and the file size is reduced to half.

```sh
> pdfsizeopt --do-require-image-optimizers=no --use-pngout=no 1.pdf output.pdf
info: prepending to PATH: /usr/bin
error: image optimizer not found on PATH: jbig2
info: loading PDF from: 1.pdf
info: loaded PDF of 6192148 bytes
info: separated to 827 objs + xref + trailer
info: parsed 827 objs
info: eliminated 208 unused objs, depth=9
info: found 0 Type1 fonts loaded
info: found 2 Type1C fonts loaded
info: will optimize image XObject 15; orig width=290 height=290 colorspace=/DeviceGray bpc=1 inv=False filter=/FlateDecode dp=0 size=587 gs_device=pngmono
info: saving PNG to psotmp.2354639.img-15.parse.png
info: written 464 bytes to PNG
info: optimizing 1 images of 587 bytes in total
info: executing image converter sam2p_np: sam2p -j:quiet -pdf:2 -c zip:1:9 -s Gray1:Indexed1:Gray2:Indexed2:Rgb1:Gray4:Indexed4:Rgb2:Gray8:Indexed8:Rgb4:Rgb8:stop -- psotmp.2354639.img-15.parse.png psotmp.2354639.img-15.sam2p-np.pdf
info: loading image from: psotmp.2354639.img-15.sam2p-np.pdf
info: loading PDF from: psotmp.2354639.img-15.sam2p-np.pdf
info: loaded PDF of 1114 bytes
info: separated to 5 objs + xref + trailer
info: parsed 5 objs
info: loaded PNG IDAT of 407 bytes
info: executing image converter sam2p_pr: sam2p -j:quiet -c zip:15:9 -- psotmp.2354639.img-15.parse.png psotmp.2354639.img-15.sam2p-pr.png
info: loading image from: psotmp.2354639.img-15.sam2p-pr.png
info: loaded PNG IDAT of 495 bytes
info: optimized image XObject 15 file_name=psotmp.2354639.img-15.sam2p-np.pdf size=586 (100%) methods=sam2p_np:586,#orig:587,parse:646,sam2p_pr:734
info: saved 1 bytes (0%) on optimizable images
info: optimized 206 streams, kept 204 #orig, 2 zip
info: eliminated 396 duplicate objs
info: compressed 0 streams, kept 0 of them uncompressed
info: saving PDF with 223 objs to: output.pdf
info: generated object stream of 2063 bytes in 114 objects (5%)
info: generated 3127680 bytes (51%)
```

Is is possible for WeasyPrint to have this optimization built-in?

**Comment by DimaBogdan-phie** [comment]
Is there any progress? 
I think the problem is not very difficult to solve.
Looks like such duplicates optimization **already implemented** in every single `document` after `HTML().render()`.
But when I try to do this: 
```python
from weasyprint import HTML
d = HTML('html.html').render()
d2 = HTML('html.html').render()
documents = [d, d2]
all_pages = [p for doc in documents for p in doc.pages]
documents[0].copy(all_pages).write_pdf('combined.pdf')
```
I get dublesized pdf (2,8mb instead of 1,4mb)

How to do such multipage optimization?

**Comment by liZe** [comment]
A new `image_cache` option can be passed to `render()` to share the cache used for images.

> I think the problem is not very difficult to solve.

I think it was!

**Comment by leesei** [comment]
Thanks.
So to use it I just need to maintain the cache in my app and pass it through various API calls?

**Comment by liZe** [comment]
> So to use it I just need to maintain the cache dict in my app and pass it through various API calls?

That’s it!

**Comment by leesei** [comment]
I updated my code to add `image_cache`, but it doesn't seems to work at all.
The generated file is still 6.2MB and `pdfsizeopt` is able to optimize half of it.
Am I doing anything wrong?

```py
html = template.render(context=context)
# html is a page of 1000 logos and 1000 QR codes (both png), totaling 200 pages

image_cache = {}
doc = HTML(string=html, base_url="./")
doc.write_pdf("1.pdf", image_cache=image_cache)
```

**Comment by liZe** [comment]
Hello!

The feature added in this issue is the possibility to store only once the same images in the PDF when it’s created using multiple `render()` calls. When you use the `HTML` class only once, these images are already de-duplicated.

`pdfsizeopt` has a lot of ways to optimize the size of PDF files, and some seem to be very efficient with your files. But trimming duplicate images, as discussed in this issue, is already done by WeasyPrint.

You can try to use the current 53 beta version, because you’ll probably get better results. If `pdfsizeopt` is still able to greatly optimize your document size, you may open a new issue with your HTML/CSS/images sample, we’ll try to find a way to generate a smaller file.

**Comment by leesei** [comment]
I see, I'll try and report later.

Any pointer to study the instructions in the generated PDF and how `pdfsizeopt` optimized them?
I can help out in studying this issue.
Thanks.

**Comment by leesei** [comment]
I've tried 53.0b2 but unfortunately it was a fail.
It took 48 seconds to create the PDF of 141MB (the CPU load was not high).
v52.5 took 13 seconds to create the PDF of 5.9MB.

**Comment by liZe** [comment]
> I've tried 53.0b2 but unfortunately it was a fail.
> It took 48 seconds to create the PDF of 141MB (the CPU load was not high).
> v52.5 took 13 seconds to create the PDF of 5.9MB.

That’s a big problem, but probably unrelated to this issue. Could you please open a new issue with a sample showing the problem? Thank you!

**Comment by leesei** [comment]
Repo to reproduce this issue: https://github.com/leesei/WeasyPrint_1392
